Failed to replicate OCI artifact to jfrog registry

When replicate image from Harbor to JFrog registry, it failed with the following error message 
```
Jul 22 11:15:43  core[834393]: 2025-07-22T11:15:43Z [ERROR] [/pkg/reg/adapter/jfrog/adapter.go:144]: Create Namespace kob-1006479-oci-dev-local error: http error: code 403, message {
Jul 22 11:15:43  core[834393]:   "errors" : [ {
Jul 22 11:15:43  core[834393]:     "status" : 403,
Jul 22 11:15:43  core[834393]:     "message" : "Forbidden"
Jul 22 11:15:43  core[834393]:   } ]
```

Only docker artifact can be sync, but the other types of OCI artifact fails.

https://github.com/goharbor/harbor/blob/abff280414be3e84e97838a0bd18f990258205df/src/pkg/reg/adapter/jfrog/adapter.go#L127-L148

In the line 127, it only fetch the type of docker, for other types it doesn't fetch, then it failed to create the repository because it is already exist.

According the API.
https://jfrog.com/help/r/jfrog-rest-apis/get-repositories-by-type-and-project
It should support other packageType such as helm, yum, debian etc. 
current packageType is hardcoded with `docker`
see https://github.com/goharbor/harbor/blob/abff280414be3e84e97838a0bd18f990258205df/src/pkg/reg/adapter/jfrog/client.go#L64

--- a/src/pkg/reg/adapter/jfrog/adapter.go
+++ b/src/pkg/reg/adapter/jfrog/adapter.go
@@ -19,6 +19,7 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"os"
 	"strings"
 
 	common_http "github.com/goharbor/harbor/src/common/http"
@@ -40,6 +41,13 @@ func init() {
 		return
 	}
 	log.Infof("the factory of jfrog artifactory adapter was registered")
+
+	// if the defaultSupportedTypes doesn't include the types, set it by env JFROG_SUPPORTED_REPOSITORY_TYPES.
+	// get supported repository types from environment variable
+	types := os.Getenv("JFROG_SUPPORTED_REPOSITORY_TYPES")
+	if types != "" {
+		supportedTypes = strings.Split(types, ",")
+	}
 }
 
 type factory struct {
@@ -56,8 +64,10 @@ func (f *factory) AdapterPattern() *model.AdapterPattern {
 }
 
 var (
-	_ adp.Adapter          = (*adapter)(nil)
-	_ adp.ArtifactRegistry = (*adapter)(nil)
+	_                     adp.Adapter          = (*adapter)(nil)
+	_                     adp.ArtifactRegistry = (*adapter)(nil)
+	defaultSupportedTypes                      = []string{"docker", "oci", "helm", "helm-oci"}
+	supportedTypes                             = defaultSupportedTypes
 )
 
 // Adapter is for images replications between harbor and jfrog artifactory image repository
@@ -124,12 +134,10 @@ func (a *adapter) PrepareForPush(resources []*model.Resource) error {
 		}
 	}
 
-	repositories, err := a.client.getDockerRepositories()
+	repositories, err := a.listAllRepositories()
 	if err != nil {
-		log.Errorf("Get local repositories error: %v", err)
 		return err
 	}
-
 	existedRepositories := make(map[string]struct{})
 	for _, repo := range repositories {
 		existedRepositories[repo.Key] = struct{}{}
@@ -139,7 +147,7 @@ func (a *adapter) PrepareForPush(resources []*model.Resource) error {
 		if _, ok := existedRepositories[namespace]; ok {
 			log.Debugf("Namespace %s already existed in remote, skip create it", namespace)
 		} else {
-			err = a.client.createDockerRepository(namespace)
+			err := a.client.createDockerRepository(namespace)
 			if err != nil {
 				log.Errorf("Create Namespace %s error: %v", namespace, err)
 				return err
@@ -201,6 +209,10 @@ func (a *adapter) FetchArtifacts(filters []*model.Filter) ([]*model.Resource, er
 	return resources, nil
 }
 
+func (a *adapter) listAllRepositories() ([]*repository, error) {
+	return a.client.getAllRepositories()
+}
+
 // listRepositories lists repositories from jfrog
 func (a *adapter) listRepositories(filters []*model.Filter) ([]*model.Repository, error) {
 	pattern := ""
@@ -210,39 +222,44 @@ func (a *adapter) listRepositories(filters []*model.Filter) ([]*model.Repository
 			break
 		}
 	}
-	var repositories []string
+	var totalRepos []string
 	// if the pattern of repository name filter is a specific repository name, just returns
 	// the parsed repositories and will check the existence later when filtering the tags
 	if paths, ok := util.IsSpecificPath(pattern); ok {
-		repositories = paths
+		totalRepos = paths
 	} else {
 		// search repositories from catalog API
-		dockerRepos, err := a.client.getDockerRepositories()
+		repositories, err := a.listAllRepositories()
 		if err != nil {
 			return nil, err
 		}
-
-		for _, docker := range dockerRepos {
-			url := fmt.Sprintf("%s/artifactory/api/docker/%s", a.client.url, docker.Key)
+		for _, r := range repositories {
+			url := buildRepositoryURL(a.client.url, r.PackageType, r.Key)
 			regClient := registry.NewClientWithAuthorizer(url, basic.NewAuthorizer(a.client.username, a.client.password), a.client.insecure, a.registry.CACertificate)
 			repos, err := regClient.Catalog()
 			if err != nil {
-				return nil, err
+				// log the error and continue to the next repository
+				log.Debugf("List catalog for repository %s error: %v", r.Key, err)
+				continue
 			}
 
 			for _, repo := range repos {
-				repositories = append(repositories, fmt.Sprintf("%s/%s", docker.Key, repo))
+				totalRepos = append(totalRepos, fmt.Sprintf("%s/%s", r.Key, repo))
 			}
 		}
 	}
 
 	var result []*model.Repository
-	for _, repository := range repositories {
+	for _, repository := range totalRepos {
 		result = append(result, &model.Repository{
 			Name: repository,
 		})
 	}
 	return filter.DoFilterRepositories(result, filters)
+}
+
+func buildRepositoryURL(baseURL, packageType, repoKey string) string {
+	return fmt.Sprintf("%s/artifactory/api/%s/%s", baseURL, strings.ToLower(packageType), repoKey)
 }
 
 // listArtifacts lists one repository tags
@@ -254,11 +271,19 @@ func (a *adapter) listArtifacts(repository string, filters []*model.Filter) ([]*
 		key = s[0]
 		repoName = strings.Join(s[1:], "/")
 	}
-	url := fmt.Sprintf("%s/artifactory/api/docker/%s", a.client.url, key)
-	regClient := registry.NewClientWithAuthorizer(url, basic.NewAuthorizer(a.client.username, a.client.password), a.client.insecure, a.registry.CACertificate)
-	tags, err := regClient.ListTags(repoName)
-	if err != nil {
-		return nil, err
+
+	// because repository can be other types like oci, helm, try to list tags with all supported types
+	var tags []string
+	for _, t := range supportedTypes {
+		url := buildRepositoryURL(a.client.url, t, key)
+		regClient := registry.NewClientWithAuthorizer(url, basic.NewAuthorizer(a.client.username, a.client.password), a.client.insecure, a.registry.CACertificate)
+		var err error
+		tags, err = regClient.ListTags(repoName)
+		if err != nil {
+			log.Debugf("List tags for repository %s with type %s error: %v", repository, t, err)
+			continue
+		}
+		break
 	}
 
 	var artifacts []*model.Artifact

--- a/src/pkg/reg/adapter/jfrog/adapter_test.go
+++ b/src/pkg/reg/adapter/jfrog/adapter_test.go
@@ -43,21 +43,65 @@ func getMockAdapter(t *testing.T, hasCred, health bool) (*adapter, *httptest.Ser
 			Pattern: "/artifactory/api/repositories",
 			Handler: func(w http.ResponseWriter, r *http.Request) {
 				w.WriteHeader(http.StatusOK)
-				w.Write([]byte(`[
-    {
-        "key": "cyzhang",
-        "description": "",
-        "type": "LOCAL",
-        "url": "http://49.4.2.82:8081/artifactory/cyzhang",
-        "packageType": "Docker"
-    },
-    {
-        "key": "mydocker",
-        "type": "LOCAL",
-        "url": "http://49.4.2.82:8081/artifactory/mydocker",
-        "packageType": "Docker"
-    }
-]`))
+				r.ParseForm()
+				typeParam := r.Form.Get("packageType")
+				// respond with different data based on the "type" query param
+				if len(typeParam) == 0 {
+					// no type param, return all repos
+					w.Write([]byte(`[
+						{
+							"key": "cyzhang",
+							"description": "",
+							"type": "LOCAL",
+							"url": "http://49.4.2.82:8081/artifactory/cyzhang",
+							"packageType": "Docker"
+						},
+						{
+							"key": "mydocker",
+							"type": "LOCAL",
+							"url": "http://49.4.2.82:8081/artifactory/mydocker",
+							"packageType": "Docker"
+						},
+						{
+							"key": "oci-repo",
+							"type": "LOCAL",
+							"url": "http://49.4.2.82:8081/artifactory/myoci",
+							"packageType": "OCI"
+						} 
+					]`))
+					return
+				} else if typeParam == "docker" {
+					// only docker repos
+					w.Write([]byte(`[
+						{
+							"key": "cyzhang",
+							"description": "",
+							"type": "LOCAL",
+							"url": "http://49.4.2.82:8081/artifactory/cyzhang",
+							"packageType": "Docker"
+						},
+						{
+							"key": "mydocker",
+							"type": "LOCAL",
+							"url": "http://49.4.2.82:8081/artifactory/mydocker",
+							"packageType": "Docker"
+						}
+					]`))
+					return
+				} else if typeParam == "oci" {
+					// only oci repos
+					w.Write([]byte(`[
+						{
+							"key": "oci-repo",
+							"type": "LOCAL",
+							"url": "http://49.4.2.82:8081/artifactory/myoci",
+							"packageType": "OCI"
+						} ]`))
+					return
+				}
+
+				w.Write([]byte(`[]`))
+
 			},
 		},
 		&test.RequestHandlerMapping{
@@ -201,4 +245,13 @@ func TestAdapter_FetchArtifacts(t *testing.T) {
 	assert.Equal(t, "mydocker/nginx", res[0].Metadata.Repository.Name)
 	assert.Len(t, res[0].Metadata.Artifacts, 1)
 	assert.Equal(t, "v1", res[0].Metadata.Artifacts[0].Tags[0])
+}
+
+func TestAdapter_ListAllRepositories(t *testing.T) {
+	a, s := getMockAdapter(t, true, true)
+	defer s.Close()
+
+	repos, err := a.listAllRepositories()
+	assert.Nil(t, err)
+	assert.Len(t, repos, 3) // 2 docker + 1 oci
 }

--- a/src/pkg/reg/adapter/jfrog/client.go
+++ b/src/pkg/reg/adapter/jfrog/client.go
@@ -61,10 +61,9 @@ func newClient(reg *model.Registry) *client {
 	}
 }
 
-// getDockerRepositories gets docker repositories from jfrog
-func (c *client) getDockerRepositories() ([]*repository, error) {
+func (c *client) getAllRepositories() ([]*repository, error) {
 	var repositories []*repository
-	url := fmt.Sprintf("%s/artifactory/api/repositories?packageType=docker", c.url)
+	url := fmt.Sprintf("%s/artifactory/api/repositories", c.url)
 	req, err := http.NewRequest(http.MethodGet, url, nil)
 	if err != nil {
 		return repositories, err

--- a/src/pkg/reg/adapter/jfrog/client_test.go
+++ b/src/pkg/reg/adapter/jfrog/client_test.go
@@ -76,13 +76,6 @@ func (c *clientTestSuite) TearDownSuite() {
 	c.mockServer.Close()
 }
 
-func (c *clientTestSuite) TestGetDockerRepositories() {
-	repos, err := c.client.getDockerRepositories()
-	c.NoError(err)
-	c.Len(repos, 2)
-	c.Equal("repo1", repos[0].Key)
-}
-
 func (c *clientTestSuite) TestCreateDockerRepository() {
 	err := c.client.createDockerRepository("test")
 	c.NoError(err)

